simpleviewinc/keg-herkin
# Zen 548 runner tab

**Ticket**: [ZEN-548](https://jira.simpleviewtools.com/browse/ZEN-548)

> **Notes**
> This only covers half the ZEN-548 ticket, which updates the Runner tab
> The rest of the ticket is being done separately on another branch
> There are already a lot of changes
> So, I figured it's best if I send the PR in parts

## Context

* <s>We want the user to be able to easily switch over the test file they are running</s>
* <s>We can either add a dropdown component on the runner header or utilize the side menu</s>
* We also need to migrate the test output container to not be connected to the test runner container
* So we have more space for the text editor
* <s>Try implementing with this editor https://github.com/SmartBear/react-gherkin-editor</s>

## Goal

* Clean up the Test runner tab
* Separate the Test results from the Test definitions/features
* Improve overall UI

## Updates

* `repos/tap/actions/*`
  * Normalized actions to be more consistent
  * Move all loadFile calls to a utils helper
  * Updated paths as needed
* `repos/tap/utils/api/*`
  * Added helpers for `load` / `save` / `delete` api actions
  * Moved `apiRequest` utils into api folder
  * Updated paths as needed
* `upsertActiveRunnerTest`
  * Updated to accept an object with a content property
  * This allows it to be consistent with a parsed `feature file` and a `jest test file`
* `repos/tap/actions/init.js`
  * Cleaned up and refactored to work with other updates and utils
* `repos/tap/assets/icons/*`
  * Added two check icons
    * Used when displaying Test Results 
  * Added a cog icon
    * Used for opening the test setting modal
* `repos/tap/components/aceEditor/aceEditor.js`
  * Lots of cleanup to allow for greater configuration
  * Added a number of setting calls for improving UI
  * Added `maxLines` setting to allow the editor to auto-grow and resize
* `repos/tap/components/aceEditor/themeOverrides.js`
  * Added a component the allows passing `CSSinJS`, and converts it to a style sheet
  * **IMPORTANT** - This should be moved to `ReTheme` at some point
    * If could reuse the style injector and expose a component that works the same as this one
* `repos/tap/components/hocs/withAppHeader.js`
  * Added Cog button to allow opening the setting modal
* `repos/tap/components/modals/testSelectorModal.js`
  * Updated action call, so when picking a feature, it gets set as active
* `repos/tap/components/runner/results.js`
  * Added check filled icon to results
  * Cleaned up lots of styles
* `repos/tap/components/runner/runner.js`
  * Refactored and moved sub-components to own files
  * Cleaned up lots of styles
  * Moved `Results` component to it's own Surface component
  * Updated `Results` to only show when the tests have been run
* `repos/tap/components/runner/testsRunning.js`
  * Pulled component from `runner.js` component
* `repos/tap/components/runner/toRun.js`
  * Cleaned up lots of styles
  * Moved some code from `runner.js` component to here
* `repos/tap/components/sidebar/sidebar.js`
  * Added toggle functionality to allow auto-closing the sidebar when clicking outside of it
* `repos/tap/components/sidebar/sidebarToggle.js`
  * Allow auto-closing the sidebar when clicking outside of it
* `repos/tap/components/surface/drawerToggle.js`
  * Moved from sub-surface component, so it can be reused
* `repos/tap/components/surface/subsurface.js`
  * Pulled drawer toggle out into it's own component
* `repos/tap/components/surface/surface.js`
  * Added header toggle functionality
  * Cleaned up lots of styles
* `repos/tap/components/tabbar/tabbar.js`
  * Updated to only show a tabs view when it actually exists
* `repos/tap/hooks/useTestRunner.js`
  * Updated to auto-toggle the Tests and Results sections based on the running state of the tests
* `repos/tap/theme/*`
  * Lots of styles cleanup
* `repos/tap/theme/components/aceEditor/themeOverrides.js`
  * Ace theme override styles
  * These get converted into a CSS style sheet and appended to the dom
  * They do not get used in the theme like other styles
  * The `keys` are `classNames` of dom elements created by the Ace Editor
* `repos/tap/utils/api/*`
  * Created a few helpers for call the backend api
  * Used by the actions to managing backend files


## Testing

**Setup**
* Pull this PR => `keg herkin && keg pr 11`
* Run the package => `keg herkin pack run keg-herkin:zen-548-runner-tab`
  * Navigate to http://herkin-zen-548-runner-tab.local.kegdev.xyz/ in a browser tab
* In your editor (VS Code), and open the file at `keg-herkin/repos/example/src/index.js`
  * Update line `8` of that file to be ```src: `${process.env.HERKIN_URL}/?tab=runner` ```
    * This will tell herkin to auto-load the `Runner` tab
    * This will be used later in **Test 4**
* In another terminal window, run `keg herkin && yarn ex:mess`
  * Navigate to http://localhost:3000/ in a different browser tab

**Test 1**
* In the **non-iframed** herkin browser tab, a modal should show up
  * [Looks like this](http://snpy.in/1FH54b)
  * Select the `google search` test from the `Select test file` drop down
  * Then click the start button
* Ensure it opens the `Core Editor` tab, and the `google search` test is loaded
  * **Note** - The definitions on the right side do not display properly
    * This is expected and will be fixed in this [ticket](https://jira.simpleviewtools.com/browse/ZEN-546)
* Click the `Feature Builder` tab, and ensure the `google search` test loads there as well

**Test 2**
* Click the `cog` in the top right, and ensure it opens the settings modal
* In the Select tag dropdown, select `Runner`, then click start
* The modal should close, and it should switch to the `Tests Runner` tab
* Ensure the `google search` feature file is loaded in the `Runner` section
  * [Looks like this](http://snpy.in/QhGqox) 

**Test 3**
* In the top right of the `Runner` section, there should be a minus icon, next to `Hide` text
* Click it, and ensure it toggles the section closed
  * It should also switch to a plus icon, next to `Show` text
  * [Looks like this](http://snpy.in/lB77ej)

**Test 4**
* Switch to the **iframed** tab, with the example app
* Toggle open the iframed herkin app from the blue arrow on the top right
  * It should auto load the `Runner` tab because we update the url in the setup section
* Click the run button at the bottom right
  * Ensure it briefly covers the `Runner` section with loading spinner
* It should then toggle the tests closed, and show the `Results` section
  * [Looks like this](http://snpy.in/CoIjxz)
  * Ensure all tests pass
  * Ensure the title of the `Resutls` section says `Tests Passed` in green text
* Toggle the `Runner` section open by clicking on the `Show` toggle in the top right of the section
* On line `23` of the `Runner` section, change the `left` text to be something that fails
  * [Looks like this](http://snpy.in/MFU1jl)
* Click the run button, and the tests should **FAIL** this time
  * Ensure the loading spinner covers the `Runner`, same as before as the tests run
* After the tests run
  * Ensure the `Runner` section is toggled closed
  * Ensure the the `Results` section is toggled open
  * Ensure the `Results` section header says `1 Failed Test(s)` in red text
  * Ensure the Failed test displays in red, and failed text info is displayed below it
  * [Looks like this](http://snpy.in/qv2nm8)
